src/Orleans.Transactions/Utilities/MultiCompletionSource.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Orleans.Transactions
+{
+    /// <summary>
+    /// Utility class that waits for the completion of multiple other tasks.
+    /// Duplication of internal orleans utility MultiTaskCompletionSource.
+    /// TODO: Consider making orleans version public.
```

Comment:
or link files, or something else, let's not make non-Orleans utilities public, as we couple runtime versions unnecessarily. See `Sharing non-core Code` in [https://github.com/dotnet/orleans/issues/3353](https://github.com/dotnet/orleans/issues/3353)

Reply:
These are not public.  They are just duplicates of useful internal Orleans logic; logic these systems used when initially developed because they were part of Orleans core.  If Orleans core does not expose and support utilities like this publically, which I agree is not necessarily a good idea, then libraries need to provide their own solutions, which is what this library is doing, it's just doing so by copying the useful internal logic it originally used.

Reply:
Correct, what I meant (as I also mentioned in person) is that instead of adding a TODO that says "consider making orleans version public", have something like `// TODO: replace with the approach that we'll implement for "Sharing non-core Code" in https://github.com/dotnet/orleans/issues/3353`